### PR TITLE
[FIRRTL] [RTL] Introduce ImplicitSSAName directive

### LINKED ISSUE #421: [FIRRTL] Remove node name attribute if the same as Value
Don't print out the `name` attribute if it's the same as the SSA value name for a node. This already happens for `wire`, but not for `node.

Currently, things print as:

```mlir
%x = firrtl.node %a {name = "x"} : !firrtl.uint<1>
```

When this could be simplified to:

```mlir
%x = firrtl.node %a : !firrtl.uint<1>
```

Similar to #90.

### PR DESCRIPTION
**Pull request by uenoku**
Will close #421. 
This PR adds Fancy custom directive by porting logic used for `wire` op. 
This PR also changes to use Fancy directive in node op and wire op

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td [hunk 1]
```diff
@@ -94,7 +94,7 @@ def NodeOp : FIRRTLOp<"node", [NoSideEffect, SameOperandsAndResultType]> {
   let arguments = (ins PassiveType:$input, OptionalAttr<StrAttr>:$name);
   let results = (outs FIRRTLType:$result);
 
-  let assemblyFormat = "$input attr-dict `:` type($input)";
+  let assemblyFormat = "$input custom<Fancy>(attr-dict) `:` type($input)";
```
Comment (lattner) [inline include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:97]: Please name this soemthing more specific than "Fancy". Maybe `custom<ImplicitSSAName>` ?

**Review by lattner — APPROVED** [review]
OMG, I love this.  Thank you so much for figuring this out!

With this, I bet a bunch of other things can adopt this (e.g. firrtl.node and firrtl.instance) and there places in the RTL dialect that can do the same thing (rtl.wire).  I love it!!!

File: include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td (hunk 1 above)
Comment (uenoku) [reply include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td]: Thanks. I changed it.

**Comment by uenoku** [comment]
> RTL dialect that can do the same thing (rtl.wire).
Sounds good. Where should I put the implementation of `parse/printImplicitSSAName`?

**Comment by lattner** [comment]
For now, I'd duplicate in both firrtl and rtl dialects, but providing it as a general MLIR thing (and replacing them when ready) makes tremendous sense to me, thanks @uenoku !

**Review by darthscsi — APPROVED** [review]
Please add a test case where the name attribute still needs to be output.

**Comment by uenoku** [comment]
There are a few problems regarding the custom directive. 

1. I wanted to apply `ImplicitSSAName` for `firrtl.instance` but it doesn't work well. 
2. The assembly format of `firrtl.instance` is as follows. 
```
  let assemblyFormat = "$moduleName attr-dict `:` type($result)";
```

`moduleName` is stored as an attribute so if I use a custom directive for `attr-dict`, $moduleName will be emitted twice. 

#### Generated printer without ImplicitSSAName
```c++
void InstanceOp::print(::mlir::OpAsmPrinter &p) {
  p << "firrtl.instance";
  p << ' ';
  p.printAttributeWithoutType(moduleNameAttr());
  p << ' ';
  printImplicitSSAName(p, *this, getOperation()->getAttrDictionary());
```

#### Generated printer with ImplicitSSAName

```c++
void InstanceOp::print(::mlir::OpAsmPrinter &p) {
  p << "firrtl.instance";
  p << ' ';
  p.printAttributeWithoutType(moduleNameAttr());
  p.printOptionalAttrDict(getAttrs(), /*elidedAttrs=*/{"moduleName"});
  p << ' ' << ":";
```

**Comment by lattner** [comment]
Meta: Please land this patch for WireOp (and anything else easy), and we can continue to iterate on it for other ops :-)

For this case, would something like this work?
`  let assemblyFormat = "$moduleName custom<ImplicitSSAName>(attr-dict, $moduleName) `:` type($result)";`

You could then define the parse/print method to take an optional extra attribute to ignore.

**Comment by uenoku** [comment]
> For this case, would something like this work?
> ` let assemblyFormat = "$moduleName custom<ImplicitSSAName>(attr-dict, $moduleName)`:` type($result)";`
Yes, I think so but we need to modify the custom directive in MLIR.

> Meta: Please land this patch for WireOp (and anything else easy), and we can continue to iterate on it for other ops :-)
OK. I'll close it. Thank you for the review!

**Comment by lattner** [comment]
great job!

**Pull request merged**